src/Symfony/Component/Translation/Translator.php

```diff
@@ -65,36 +52,74 @@ class Translator implements TranslatorInterface, TranslatorBagInterface
      */
-    private $configCacheFactory;
+    private $messageCatalogueProvider;
 
     /**
-     * Constructor.
-     *
-     * @param string               $locale   The locale
-     * @param MessageSelector|null $selector The message selector for pluralization
-     * @param string|null          $cacheDir The directory to use for the cache
-     * @param bool                 $debug    Use cache in debug mode ?
+     * @var ResourceMessageCatalogueProvider
+     */
+    private $resourceMessageCatalogueProvider;
+
+    /**
+     * @var CacheMessageCatalogueProvider
+     */
+    private $cacheMessageCatalogueProvider;
+
+    /**
+     * @param string                                                 $locale                   The locale
+     * @param MessageCatalogueProviderInterface|MessageSelector|null $messageCatalogueProvider The MessageCatalogueProviderInterface or MessageSelector
+     *                                                                                         Passing the MessageSelector or null as a second parameter is deprecated since version 2.8.
+     * @param MessageSelector|string|null                            $selector                 The MessageSelector or cache directory
+     *                                                                                         Passing the cache directory as a third parameter is deprecated since version 2.8.
+     * @param bool                                                   $debug                    Use cache in debug mode ?
+     *                                                                                         Deprecated since version 2.8, to be removed in 3.0.
      *
      * @throws \InvalidArgumentException If a locale contains invalid characters
      */
-    public function __construct($locale, MessageSelector $selector = null, $cacheDir = null, $debug = false)
+    public function __construct($locale, $messageCatalogueProvider = null, $selector = null, $debug = false)
     {
         $this->setLocale($locale);
-        $this->selector = $selector ?: new MessageSelector();
-        $this->cacheDir = $cacheDir;
-        $this->debug = $debug;
+
+        if ($messageCatalogueProvider instanceof MessageCatalogueProviderInterface) {
+            $this->messageCatalogueProvider = $messageCatalogueProvider;
+            $this->selector = $selector ?: new MessageSelector();
+
+            do {
+                if ($messageCatalogueProvider instanceof CacheMessageCatalogueProvider) {
+                    $this->cacheMessageCatalogueProvider = $messageCatalogueProvider;
+                } elseif ($messageCatalogueProvider instanceof ResourceMessageCatalogueProvider) {
+                    $this->resourceMessageCatalogueProvider = $messageCatalogueProvider;
+                }
+
+                $messageCatalogueProvider = method_exists($messageCatalogueProvider, 'getInner') ? $messageCatalogueProvider->getInner() : null;
+            } while ($messageCatalogueProvider);
```

Comment:
this allow to use the legacy methods even if we use the new way but if we assume it shouldn't be used in that case we can remove the entire block. what do you  think ?